{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Challenge Problem 3: Genome Wide Association Study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background:\n",
    "\n",
    "(Adapted From BMI 704, Dr. Chirag Patel)\n",
    "\n",
    "Genome-wide association studies (GWASs) are arguably the most significant biomedical advance in the last decade. Utilizing an observational case-control design, populations with disease are analytically compared to those without to discern patterns in differences in genetic variant, or single nucleotide polymorphism (SNP),frequencies in each of the respective populations. If statistically differences in SNP frequencies in cases vs\n",
    "controls are found, one can conclude the SNP locus may have a causal relationship, implicating a gene, genetic region, or gene regulator in the disease of interest.\n",
    "\n",
    "Analyzing GWASs provide lessons in observational research, including statistical association and multiple testing. They also provide a way to create hypotheses about biological basis of disease for further study. We will execute a GWAS analysis using data from the Wellcome Trust Case-Control Consortium (WTCCC) on Type 2 Diabetes, Type 1 Diabetes, Bipolar Disorder, Rheumatoid Arthritis, Chron's Disease and Hypertension versus 1,500 controls.\n",
    "\n",
    "Note: There are standard tools and pipelines available to do this but we will do it the old fashioned way!\n",
    "\n",
    "The Data is provided in separate directories under the primary WTCCC directory. The subdirectories are labeled as such:\n",
    "* 58C : control population 1\n",
    "* NBS : control population 2\n",
    "* RA : rheumatoid arthritis\n",
    "* T2D : type 2 diabetes\n",
    "* T1D : type 1 diabetes\n",
    "* CAD : coronary artery disease\n",
    "* HT : hypertension\n",
    "* CD : Crohn’s disease\n",
    "* BD : bipolar disorder\n",
    "\n",
    "All of the GWAS data is broken down per chromosome, 1-22 plus X and is written into the file name, as well as the disease. For example:\n",
    "\n",
    "Affx_gt_ 58C _Chiamo_ 06 .tped.gz\n",
    "\n",
    "Contains GWAS data for the 6th (06) chromosome on control population 1 (58C). The biggest files are chromosome 1 and 2 and smallest is 22.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row is a SNP (there are 6207 SNPs measured in Chromosome 22). The first through fourth columns denote the chromosome number, the second is the WTCCC snp_id (not to be confused with the rsid), the start coordinate (0), and the end coordinate. Thereafter, each column is a genotype (where each base is separated by a space character) for that SNP for each individual.\n",
    "\n",
    "Additionally, each subdirectory also contains a ‘snps_info.tar.gz’, which contains the snp_id and rsID for each SNP for each chromosome. These files can enable you to map the WTCCC snp_id to the rsid, which is helpful to find the closest genes to a SNP (for example when you query the GWAS catalog)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from os import path\n",
    "import csv\n",
    "import numpy as np\n",
    "from time import time\n",
    "from scipy.stats import chisquare\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "######## CLASS NOTES ON WHAT TO DO ##################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the test data \n",
    "t2d_22_file = 'BWSI_100set/T2D/Affx_gt_T2D_Chiamo_22.tped.gz' ##########THE OG CODE SAID BWSI_test INSTEAD OF BWSI_100set\n",
    "t2d = pd.read_csv(t2d_22_file, compression='gzip', header=None, sep='\\t')\n",
    "c1 = pd.read_csv('BWSI_100set/58C/Affx_gt_58C_Chiamo_22.tped.gz', compression='gzip', header=None,sep='\\t')\n",
    "c2 = pd.read_csv('BWSI_100set/NBS/Affx_gt_NBS_Chiamo_22.tped.gz', compression='gzip', header=None,sep='\\t')\n",
    "snp = pd.read_csv('BWSI_100set/58C/snps/snps_22',header=None, sep='\\t')\n",
    "t2d_head = t2d.loc[:,0:3]\n",
    "t2d_snps = t2d.loc[:,4:]\n",
    "c1_head = c1.loc[:,0:3]\n",
    "c1_snps = c1.loc[:,4:]\n",
    "c2_head = c2.loc[:,0:3]\n",
    "c2_snps = c2.loc[:,4:]\n",
    "snp = snp.drop([0,1],axis=1)\n",
    "snp = snp.rename(columns={3:'snp_name',4:'rs'} )\n",
    "c1_head = c1_head.rename(columns={1:'snp_name'})\n",
    "snp=snp.drop([2],axis=1)\n",
    "snp_c1_joined = c1_head.join(snp.set_index('snp_name'), on='snp_name')\n",
    "t2d_1 = t2d_snps.loc[0,:]\n",
    "c1_1 = c1_snps.loc[0,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [],
   "source": [
    "##1 compute Allele Frequency for the minor allele\n",
    "def computeAlleleFrequency (f_C, f_T):\n",
    "    \"\"\"3\n",
    "    f_C = minor allele count\n",
    "    f_T = major allele count\n",
    "    minor_allele_frequency = f_C/ (f_C+f_T) \n",
    "    7\"\"\"\n",
    "    minor_allele_frequency = f_C/(f_C+f_T)\n",
    "    return minor_allele_frequency\n",
    "\n",
    "def getAlleleCounts (genotype):\n",
    "    \"\"\" 12\n",
    "    genotype = \"C C \tC T \tC T \tC T \tT T\"\n",
    "    Allele frequency:\n",
    "    count the \"Cs\", and the \"Ts\"\n",
    "    f_C = # of Cs\n",
    "    f_T = # of T's\n",
    "    18\"\"\"\n",
    "    dicto = {}\n",
    "    stro = genotype.to_string()\n",
    "    if stro.count('C')!=0:\n",
    "        dicto.update({'C': stro.count('C')})\n",
    "    if stro.count('T')!=0:\n",
    "        dicto.update({'T': stro.count('T')})\n",
    "    if stro.count('G')!=0:\n",
    "        dicto.update({'G': stro.count('G')})\n",
    "    if stro.count('A')!=0:\n",
    "        dicto.update({'A': stro.count('A')})\n",
    "    if len(dicto)==1:\n",
    "        dicto.update({'-1':0})\n",
    "    return dicto\n",
    "\n",
    "def getGenotypeCounts (genotype):\n",
    "    stro = genotype.to_string()\n",
    "    cc = stro.count('C C')\n",
    "    tt = stro.count('T T')\n",
    "    aa = stro.count('A A')\n",
    "    ag = stro.count('A G')\n",
    "    gg = stro.count('G G')\n",
    "    if cc==0 and tt==0:\n",
    "        return np.array([aa,ag+stro.count('G A'),gg])\n",
    "    elif gg==0 and tt==0: \n",
    "        return np.array([aa,stro.count('A C')+stro.count('C A'),cc])\n",
    "    elif gg==0 and aa==0: \n",
    "        return np.array([cc,stro.count('C T')+stro.count('T C'),tt])\n",
    "    else:\n",
    "        return np.array([tt,stro.count('G T')+stro.count('T G'),gg])\n",
    "\n",
    "#30 # Compute the Odds Ratio\n",
    "#31 # takes as input a 2X2  confusion matrix\n",
    "#32 # returns the odds ratio\n",
    "def computeOR(health,diabetic):\n",
    "    if health[0] ==0 or diabetic[1]==0:\n",
    "        return 0\n",
    "    else:\n",
    "        return diabetic[0]*health[1]/(health[0]*diabetic[1])\n",
    "\n",
    "#36 # Execute a chisq test to determine if the odds ratio is significant\n",
    "#37 # return the p-value from the chisq test\n",
    "def getPValue(confusionmatrix):\n",
    "    return (scipy.stats.chisquare(confusionmatrix).pvalue)\n",
    "\n",
    "def getPValue(confusionMatrix):\n",
    "    from scipy.stats import chisquare\n",
    "    return chisquare(confusionMatrix).pvalue\n",
    "\n",
    "def get_snp_stats(control, disease):\n",
    "    \n",
    "    health_v = list(getAlleleCounts(control).values())\n",
    "    health_k = list(getAlleleCounts(control))\n",
    "    diabetic_v = list(getAlleleCounts(disease).values())\n",
    "    diabetic_k = list(getAlleleCounts(disease))\n",
    "    \n",
    "# 59    dis_freq = computeAlleleFrequency(diabetic[0], diabetic[1])\n",
    "# 60    health_freq =  computeAlleleFrequency(health[0], health[1])\n",
    "\n",
    "    o_r= computeOR(health_v,diabetic_v) # odds ratio\n",
    "    \n",
    "    \n",
    "    minor_allele_con = health_k[0]\n",
    "    major_allele_con = health_k[1]\n",
    "    minor_allele_dis = diabetic_k[0]\n",
    "    major_allele_dis = diabetic_k[1]\n",
    "\n",
    "    t1_minor = computeAlleleFrequency(diabetic_v[0], diabetic_v[1])\n",
    "    t1_major = 1 - t1_minor\n",
    "    control_minor = computeAlleleFrequency(health_v[0], health_v[1])\n",
    "    control_major = 1-control_minor\n",
    "    \n",
    "    \n",
    "    p = getPValue([t1_minor, t1_major,control_minor, control_major])\n",
    "\n",
    "    if control.to_string().count('C C')!=0 and control.to_string().count('C T')!=0 and control.to_string().count('T T')!=0:\n",
    "        HWE = HWEChiSq(control_minor, getGenotypeCounts(control))\n",
    "    else:\n",
    "        HWE = HWEChiSq(control_minor, getGenotypeCounts(control))\n",
    "    \n",
    "    \n",
    "    result=  pd.DataFrame(np.array([minor_allele_con,major_allele_con,minor_allele_dis,major_allele_dis,control_minor,t1_minor,o_r,p,HWE]))\n",
    "    return result\n",
    "    \n",
    "## function to load the data file\n",
    "def loadFile (disease, chrom_num):\n",
    "    ## construct your file name based on the disease and chrom_number\n",
    "    if chrom_num<10:\n",
    "        stro = '0'+str(chrom_num)\n",
    "    else:\n",
    "        stro = str(chrom_num)\n",
    "    fileName = 'BWSI_100set/'+disease+'/Affx_gt_'+disease+'_Chiamo_'+stro+'.tped.gz'\n",
    "    dat = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "    return (dat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>snp_name</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>rs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP_A-1938722</td>\n",
       "      <td>0</td>\n",
       "      <td>15385344</td>\n",
       "      <td>rs761296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP_A-4217222</td>\n",
       "      <td>0</td>\n",
       "      <td>243847383</td>\n",
       "      <td>rs11582843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP_A-2023862</td>\n",
       "      <td>0</td>\n",
       "      <td>192847341</td>\n",
       "      <td>rs3009323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP_A-4196224</td>\n",
       "      <td>0</td>\n",
       "      <td>95890895</td>\n",
       "      <td>rs17114028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP_A-2286934</td>\n",
       "      <td>0</td>\n",
       "      <td>189084167</td>\n",
       "      <td>rs16832860</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0       snp_name  2          3          rs\n",
       "0  1  SNP_A-1938722  0   15385344    rs761296\n",
       "1  1  SNP_A-4217222  0  243847383  rs11582843\n",
       "2  1  SNP_A-2023862  0  192847341   rs3009323\n",
       "3  1  SNP_A-4196224  0   95890895  rs17114028\n",
       "4  1  SNP_A-2286934  0  189084167  rs16832860"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_c1_joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Chromosome 1...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:31: RuntimeWarning: invalid value encountered in double_scalars\n",
      "/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:36: RuntimeWarning: divide by zero encountered in double_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Chromosome 2...\n",
      "Doing Chromosome 3...\n",
      "Doing Chromosome 4...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:36: RuntimeWarning: invalid value encountered in double_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Chromosome 5...\n",
      "Doing Chromosome 6...\n",
      "Doing Chromosome 7...\n",
      "Doing Chromosome 8...\n",
      "Doing Chromosome 9...\n",
      "Doing Chromosome 10...\n",
      "Doing Chromosome 11...\n",
      "Doing Chromosome 12...\n",
      "Doing Chromosome 13...\n",
      "Doing Chromosome 14...\n",
      "Doing Chromosome 15...\n",
      "Doing Chromosome 16...\n",
      "Doing Chromosome 17...\n",
      "Doing Chromosome 18...\n",
      "Doing Chromosome 19...\n",
      "Doing Chromosome 20...\n",
      "Doing Chromosome 21...\n",
      "Doing Chromosome 22...\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "bigdf = pd.DataFrame()\n",
    "snpsid = pd.DataFrame()\n",
    "for i in range(22):\n",
    "    \n",
    "    now = time.time()\n",
    "    i = i+1\n",
    "    print(\"Doing Chromosome \"+str(i)+\"...\")\n",
    "    t2d = loadFile('T2D',i)\n",
    "    c1 = loadFile('58C',i)\n",
    "    \n",
    "    if i<10:\n",
    "        stro = '0'+str(i)\n",
    "    else:\n",
    "        stro = str(i)\n",
    "    snp = pd.read_csv('BWSI_100set/58C/snps/snps_'+stro,header=None, sep='\\t')\n",
    "    snp = snp.iloc[:100,:]\n",
    "    t2d_head = t2d.loc[:,0:3]\n",
    "    t2d_snps = t2d.loc[:,4:]\n",
    "    c1_head = c1.loc[:,0:3]\n",
    "    c1_snps = c1.loc[:,4:]\n",
    "    snp = snp.drop([0,1],axis=1)\n",
    "    snp = snp.rename(columns={3:'snp_name',4:'rs'} )\n",
    "    c1_head = c1_head.rename(columns={1:'snp_name'})\n",
    "    snp=snp.drop([2],axis=1)\n",
    "    df = pd.DataFrame()\n",
    "    snpsid = snpsid.append(snp)\n",
    "    joined_con=c1_head.join(snp.set_index('snp_name','rs'), on='snp_name')\n",
    "    snpside = snpsid.append(joined_con)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    for j in range(100):\n",
    "        t2d_1 = t2d_snps.loc[j,:]\n",
    "        c1_1 = c1_snps.loc[j,:]\n",
    "        df=df.append(get_snp_stats(c1_1,t2d_1).transpose())\n",
    "#         snpsid = snpsid.append(pd.DataFrame(snp.iloc[j,0]))\n",
    "\n",
    "    bigdf=bigdf.append(df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2200, 2)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2200, 9)"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(snpsid.shape)\n",
    "bigdf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   0  1  2  3                   4                    5                   6  \\\n",
      "0  C  T  C  T  0.5309175531914894   0.5582791395697849  1.1166719025125678   \n",
      "1  C  T  C  T  0.3490691489361702  0.33616808404202103  0.9443255463451394   \n",
      "2  T  A  T  A  0.9976728723404256   0.9974987493746873  0.9302232589136954   \n",
      "3  G  A  G  A  0.9079122340425532   0.9092046023011505      1.015677563895   \n",
      "4  G  A  G  A  0.4557845744680851  0.45947973986993496  1.0149989992341526   \n",
      "\n",
      "                    7                      8  \n",
      "0  0.9993922459284262     0.9226596112930598  \n",
      "1  0.9778336483290864    0.14828324187274194  \n",
      "2  0.5764152215010026                    0.0  \n",
      "3  0.7207533541801461                    0.0  \n",
      "4   0.999542988917936  2.052818933652854e-22  \n",
      "        snp_name        rs\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n"
     ]
    }
   ],
   "source": [
    "# bigdf.reset_index(inplace=True,drop=True)\n",
    "# snpsid.reset_index(inplace=True,drop=True)\n",
    "print(bigdf.head())\n",
    "print(snpsid.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pd.concat([snpsid,bigdf],axis=1)\n",
    "a = a.rename(columns={0: 'con min all', 1: \"dis min all\",2:'con maj all',3:'dis maj all',4:'con maj all freq',5:'dis maj all freq',6:'odds rat',7:'p',8:'HWE'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [],
   "source": [
    "a.to_csv('gene_dat.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_name</th>\n",
       "      <th>rs</th>\n",
       "      <th>con min all</th>\n",
       "      <th>dis min all</th>\n",
       "      <th>con maj all</th>\n",
       "      <th>dis maj all</th>\n",
       "      <th>con maj all freq</th>\n",
       "      <th>dis maj all freq</th>\n",
       "      <th>odds rat</th>\n",
       "      <th>p</th>\n",
       "      <th>HWE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP_A-4221087</td>\n",
       "      <td>rs307378</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.5309175531914894</td>\n",
       "      <td>0.5582791395697849</td>\n",
       "      <td>1.1166719025125678</td>\n",
       "      <td>0.9993922459284262</td>\n",
       "      <td>0.9226596112930598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP_A-2203152</td>\n",
       "      <td>rs474731</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.3490691489361702</td>\n",
       "      <td>0.33616808404202103</td>\n",
       "      <td>0.9443255463451394</td>\n",
       "      <td>0.9778336483290864</td>\n",
       "      <td>0.14828324187274194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP_A-4198280</td>\n",
       "      <td>rs485777</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9976728723404256</td>\n",
       "      <td>0.9974987493746873</td>\n",
       "      <td>0.9302232589136954</td>\n",
       "      <td>0.5764152215010026</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP_A-1861214</td>\n",
       "      <td>rs492861</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9079122340425532</td>\n",
       "      <td>0.9092046023011505</td>\n",
       "      <td>1.015677563895</td>\n",
       "      <td>0.7207533541801461</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNP_A-1860431</td>\n",
       "      <td>rs496723</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.4557845744680851</td>\n",
       "      <td>0.45947973986993496</td>\n",
       "      <td>1.0149989992341526</td>\n",
       "      <td>0.999542988917936</td>\n",
       "      <td>2.052818933652854e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SNP_A-1956539</td>\n",
       "      <td>rs510688</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996343085106383</td>\n",
       "      <td>0.9979989994997499</td>\n",
       "      <td>1.8305805805805806</td>\n",
       "      <td>0.5771027456346043</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SNP_A-4255384</td>\n",
       "      <td>rs518058</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.46708776595744683</td>\n",
       "      <td>0.4652326163081541</td>\n",
       "      <td>0.992572963723705</td>\n",
       "      <td>0.9997671733953388</td>\n",
       "      <td>2.2118958676356859e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SNP_A-1963203</td>\n",
       "      <td>rs520764</td>\n",
       "      <td>G</td>\n",
       "      <td>-1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9989994997498749</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.573236918288303</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SNP_A-2178414</td>\n",
       "      <td>rs524965</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.3703457446808511</td>\n",
       "      <td>0.37018509254627313</td>\n",
       "      <td>0.9993112427149301</td>\n",
       "      <td>0.9873774881029727</td>\n",
       "      <td>2.55662516087757e-199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SNP_A-2000312</td>\n",
       "      <td>rs532419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.46043882978723405</td>\n",
       "      <td>0.46823411705852924</td>\n",
       "      <td>1.0318375553148063</td>\n",
       "      <td>0.9997229741263883</td>\n",
       "      <td>1.907260458390498e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SNP_A-4285953</td>\n",
       "      <td>rs573378</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.24900265957446807</td>\n",
       "      <td>0.24687343671835918</td>\n",
       "      <td>0.988645992730704</td>\n",
       "      <td>0.9170662006167895</td>\n",
       "      <td>0.8094762728412335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SNP_A-2062285</td>\n",
       "      <td>rs580771</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7101063829787234</td>\n",
       "      <td>0.6935967983991996</td>\n",
       "      <td>0.9241213788886341</td>\n",
       "      <td>0.954970332189847</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SNP_A-4284705</td>\n",
       "      <td>rs597479</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0003324468085106383</td>\n",
       "      <td>0.0007503751875937969</td>\n",
       "      <td>2.258072590738423</td>\n",
       "      <td>0.5733056062430879</td>\n",
       "      <td>0.9897099256674965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SNP_A-4260954</td>\n",
       "      <td>rs609031</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7752659574468085</td>\n",
       "      <td>0.7661330665332666</td>\n",
       "      <td>0.94962805331083</td>\n",
       "      <td>0.8995402824397165</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SNP_A-2073413</td>\n",
       "      <td>rs624969</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9268617021276596</td>\n",
       "      <td>0.9189594797398699</td>\n",
       "      <td>0.8947960394636626</td>\n",
       "      <td>0.6982957566932599</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SNP_A-4292038</td>\n",
       "      <td>rs666111</td>\n",
       "      <td>C</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9994997498749375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5728219156239178</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SNP_A-2253696</td>\n",
       "      <td>rs783883</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7952127659574468</td>\n",
       "      <td>0.8126563281640821</td>\n",
       "      <td>1.117088113024724</td>\n",
       "      <td>0.8638481072473968</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SNP_A-2219063</td>\n",
       "      <td>rs841392</td>\n",
       "      <td>G</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0007503751875937969</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5730294430626224</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SNP_A-2144578</td>\n",
       "      <td>rs864560</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1535904255319149</td>\n",
       "      <td>0.15607803901950976</td>\n",
       "      <td>1.0191918336553785</td>\n",
       "      <td>0.8125915710165198</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SNP_A-2190868</td>\n",
       "      <td>rs948620</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.24335106382978725</td>\n",
       "      <td>0.23936968484242122</td>\n",
       "      <td>0.9784906819909327</td>\n",
       "      <td>0.9110956219430866</td>\n",
       "      <td>0.1257407749820825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SNP_A-4264596</td>\n",
       "      <td>rs962923</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.7662898936170213</td>\n",
       "      <td>0.7638819409704852</td>\n",
       "      <td>0.9866916063090555</td>\n",
       "      <td>0.9050340372238652</td>\n",
       "      <td>0.7569943543112758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SNP_A-1786988</td>\n",
       "      <td>rs1359759</td>\n",
       "      <td>G</td>\n",
       "      <td>-1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9997498749374687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5726143360630414</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SNP_A-2275684</td>\n",
       "      <td>rs1409806</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.578125</td>\n",
       "      <td>0.5940470235117559</td>\n",
       "      <td>1.0678423340160863</td>\n",
       "      <td>0.9961803640375901</td>\n",
       "      <td>1.9425547092644167e-70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SNP_A-1857432</td>\n",
       "      <td>rs1543594</td>\n",
       "      <td>C</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9989994997498749</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.573236918288303</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SNP_A-2105952</td>\n",
       "      <td>rs1551623</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.48570478723404253</td>\n",
       "      <td>0.4912456228114057</td>\n",
       "      <td>1.0224230517883186</td>\n",
       "      <td>0.9999899814742175</td>\n",
       "      <td>0.3113572840262145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SNP_A-2213809</td>\n",
       "      <td>rs1551624</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.5575132978723404</td>\n",
       "      <td>0.545272636318159</td>\n",
       "      <td>0.9517165192965331</td>\n",
       "      <td>0.9991710115097504</td>\n",
       "      <td>3.57792661570284e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SNP_A-2118217</td>\n",
       "      <td>rs1695824</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.46675531914893614</td>\n",
       "      <td>0.4627313656828414</td>\n",
       "      <td>0.9839538005273574</td>\n",
       "      <td>0.9997357617006344</td>\n",
       "      <td>2.0239515589746665e-232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SNP_A-4262266</td>\n",
       "      <td>rs1698597</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.5428856382978723</td>\n",
       "      <td>0.5327663831915957</td>\n",
       "      <td>0.9601061353691462</td>\n",
       "      <td>0.9996666807695281</td>\n",
       "      <td>1.915509712377738e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SNP_A-2030917</td>\n",
       "      <td>rs1778121</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.840093085106383</td>\n",
       "      <td>0.8529264632316158</td>\n",
       "      <td>1.103867348284783</td>\n",
       "      <td>0.810716238260011</td>\n",
       "      <td>5.158118023773139e-292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SNP_A-4253178</td>\n",
       "      <td>rs1835283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6615691489361702</td>\n",
       "      <td>0.6745872936468235</td>\n",
       "      <td>1.0604699129776476</td>\n",
       "      <td>0.9732289082335358</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2170</th>\n",
       "      <td>SNP_A-2300067</td>\n",
       "      <td>rs41517345</td>\n",
       "      <td>C</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9979989994997499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5740662952450257</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171</th>\n",
       "      <td>SNP_A-1928576</td>\n",
       "      <td>rs41529146</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6944813829787234</td>\n",
       "      <td>0.7106053026513257</td>\n",
       "      <td>1.080226796079228</td>\n",
       "      <td>0.9545414087070017</td>\n",
       "      <td>0.4377155488291514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2172</th>\n",
       "      <td>SNP_A-4238377</td>\n",
       "      <td>rs41529846</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6781914893617021</td>\n",
       "      <td>0.6835917958979489</td>\n",
       "      <td>1.0251662404092072</td>\n",
       "      <td>0.9670392672308434</td>\n",
       "      <td>0.14798543581212498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2173</th>\n",
       "      <td>SNP_A-4280966</td>\n",
       "      <td>rs41531149</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.10871010638297872</td>\n",
       "      <td>0.11530765382691345</td>\n",
       "      <td>1.0685994022128673</td>\n",
       "      <td>0.7519530577485716</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2174</th>\n",
       "      <td>SNP_A-2314782</td>\n",
       "      <td>rs140378</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.7928856382978723</td>\n",
       "      <td>0.7781390695347674</td>\n",
       "      <td>0.9161699744031538</td>\n",
       "      <td>0.8842983002519721</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2175</th>\n",
       "      <td>SNP_A-4290489</td>\n",
       "      <td>rs5748616</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3394281914893617</td>\n",
       "      <td>0.3129064532266133</td>\n",
       "      <td>0.886279645464895</td>\n",
       "      <td>0.9703401170148667</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2176</th>\n",
       "      <td>SNP_A-2151351</td>\n",
       "      <td>rs41439550</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6509308510638298</td>\n",
       "      <td>0.6558279139569785</td>\n",
       "      <td>1.0218587429032948</td>\n",
       "      <td>0.9794635080956823</td>\n",
       "      <td>0.13234237394837803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2177</th>\n",
       "      <td>SNP_A-4284341</td>\n",
       "      <td>rs4010550</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6894946808510638</td>\n",
       "      <td>0.6835917958979489</td>\n",
       "      <td>0.9729426248565908</td>\n",
       "      <td>0.9640279235175554</td>\n",
       "      <td>0.0056153861081748215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2178</th>\n",
       "      <td>SNP_A-1854151</td>\n",
       "      <td>rs5747360</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.6891622340425532</td>\n",
       "      <td>0.6880940470235117</td>\n",
       "      <td>0.9950306205225392</td>\n",
       "      <td>0.9628896422777566</td>\n",
       "      <td>1.2680434459713623e-286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2179</th>\n",
       "      <td>SNP_A-2276339</td>\n",
       "      <td>rs2379981</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9820478723404256</td>\n",
       "      <td>0.9814907453726863</td>\n",
       "      <td>0.9693498508664385</td>\n",
       "      <td>0.6026503005703618</td>\n",
       "      <td>0.45104726180714083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2180</th>\n",
       "      <td>SNP_A-4285196</td>\n",
       "      <td>rs4535153</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.854720744680851</td>\n",
       "      <td>0.8716858429214608</td>\n",
       "      <td>1.154688332829134</td>\n",
       "      <td>0.7877271184095397</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2181</th>\n",
       "      <td>SNP_A-4229849</td>\n",
       "      <td>rs11089243</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0581781914893617</td>\n",
       "      <td>0.05502751375687844</td>\n",
       "      <td>0.9426907660893897</td>\n",
       "      <td>0.6655664382423445</td>\n",
       "      <td>7.788837202109299e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2182</th>\n",
       "      <td>SNP_A-2113726</td>\n",
       "      <td>rs5746647</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0006648936170212766</td>\n",
       "      <td>0.0005002501250625312</td>\n",
       "      <td>0.7522522522522522</td>\n",
       "      <td>0.5733740121353723</td>\n",
       "      <td>0.15736831673292917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2183</th>\n",
       "      <td>SNP_A-2214998</td>\n",
       "      <td>rs9605923</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.22340425531914893</td>\n",
       "      <td>0.2221110555277639</td>\n",
       "      <td>0.9925585668350941</td>\n",
       "      <td>0.8930107734743976</td>\n",
       "      <td>0.05452673337548858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2184</th>\n",
       "      <td>SNP_A-2296864</td>\n",
       "      <td>rs5747999</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5199468085106383</td>\n",
       "      <td>0.5500250125062531</td>\n",
       "      <td>1.1285596288930053</td>\n",
       "      <td>0.9996688083013009</td>\n",
       "      <td>2.890087559459401e-245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2185</th>\n",
       "      <td>SNP_A-2145794</td>\n",
       "      <td>rs11089263</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.14760638297872342</td>\n",
       "      <td>0.13731865932966483</td>\n",
       "      <td>0.9192088580317053</td>\n",
       "      <td>0.7957168575997363</td>\n",
       "      <td>0.7170055228220931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186</th>\n",
       "      <td>SNP_A-4295548</td>\n",
       "      <td>rs11089264</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.08078457446808511</td>\n",
       "      <td>0.07578789394697348</td>\n",
       "      <td>0.9330760620792194</td>\n",
       "      <td>0.7002014465314452</td>\n",
       "      <td>0.7770893626472968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2187</th>\n",
       "      <td>SNP_A-2256115</td>\n",
       "      <td>rs2154615</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.991688829787234</td>\n",
       "      <td>0.9947473736868434</td>\n",
       "      <td>1.5871685583385216</td>\n",
       "      <td>0.5836622508744744</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2188</th>\n",
       "      <td>SNP_A-1813743</td>\n",
       "      <td>rs5993628</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1612367021276596</td>\n",
       "      <td>0.14907453726863432</td>\n",
       "      <td>0.91135475190458</td>\n",
       "      <td>0.8129520473824803</td>\n",
       "      <td>3.0506929774379464e-99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2189</th>\n",
       "      <td>SNP_A-1792553</td>\n",
       "      <td>rs361995</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2237367021276596</td>\n",
       "      <td>0.23386693346673337</td>\n",
       "      <td>1.0590986994754534</td>\n",
       "      <td>0.8990387364709654</td>\n",
       "      <td>0.6871458183840945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2190</th>\n",
       "      <td>SNP_A-2049082</td>\n",
       "      <td>rs12106650</td>\n",
       "      <td>T</td>\n",
       "      <td>-1</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9997498749374687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5726143360630414</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2191</th>\n",
       "      <td>SNP_A-2287278</td>\n",
       "      <td>rs2845362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3796542553191489</td>\n",
       "      <td>0.38594297148574286</td>\n",
       "      <td>1.026975221232625</td>\n",
       "      <td>0.9906149141919861</td>\n",
       "      <td>1.6049536406440105e-178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2192</th>\n",
       "      <td>SNP_A-1860833</td>\n",
       "      <td>rs2845377</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3617021276595745</td>\n",
       "      <td>0.35542771385692845</td>\n",
       "      <td>0.9730877217010203</td>\n",
       "      <td>0.9837564537908076</td>\n",
       "      <td>2.277661307635628e-241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2193</th>\n",
       "      <td>SNP_A-2027961</td>\n",
       "      <td>rs2215075</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.05053191489361702</td>\n",
       "      <td>0.05127563781890945</td>\n",
       "      <td>1.0155133417514257</td>\n",
       "      <td>0.6563325720104201</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2194</th>\n",
       "      <td>SNP_A-1970657</td>\n",
       "      <td>rs17432784</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0425531914893617</td>\n",
       "      <td>0.04452226113056528</td>\n",
       "      <td>1.0484293193717278</td>\n",
       "      <td>0.6443242463857755</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2195</th>\n",
       "      <td>SNP_A-1936806</td>\n",
       "      <td>rs16984825</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9996675531914894</td>\n",
       "      <td>0.9997498749374687</td>\n",
       "      <td>1.3292317924842034</td>\n",
       "      <td>0.5728903614495753</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2196</th>\n",
       "      <td>SNP_A-2290590</td>\n",
       "      <td>rs8139954</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.041555851063829786</td>\n",
       "      <td>0.04302151075537769</td>\n",
       "      <td>1.036855201254574</td>\n",
       "      <td>0.6422823228744252</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2197</th>\n",
       "      <td>SNP_A-4198761</td>\n",
       "      <td>rs5748617</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0003324468085106383</td>\n",
       "      <td>0.0005002501250625312</td>\n",
       "      <td>1.505005005005005</td>\n",
       "      <td>0.5730980099333562</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2198</th>\n",
       "      <td>SNP_A-2078648</td>\n",
       "      <td>rs9605146</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.10272606382978723</td>\n",
       "      <td>0.09954977488744372</td>\n",
       "      <td>0.9656616325062927</td>\n",
       "      <td>0.7356143142155831</td>\n",
       "      <td>0.600652738017438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2199</th>\n",
       "      <td>SNP_A-1827561</td>\n",
       "      <td>rs16981694</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.11402925531914894</td>\n",
       "      <td>0.11480740370185093</td>\n",
       "      <td>1.0077091830556144</td>\n",
       "      <td>0.7555508549056137</td>\n",
       "      <td>0.057437616731301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2200 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           snp_name          rs con min all dis min all con maj all dis maj all       con maj all freq       dis maj all freq            odds rat                   p                      HWE\n",
       "0     SNP_A-4221087    rs307378           C           T           C           T     0.5309175531914894     0.5582791395697849  1.1166719025125678  0.9993922459284262       0.9226596112930598\n",
       "1     SNP_A-2203152    rs474731           C           T           C           T     0.3490691489361702    0.33616808404202103  0.9443255463451394  0.9778336483290864      0.14828324187274194\n",
       "2     SNP_A-4198280    rs485777           T           A           T           A     0.9976728723404256     0.9974987493746873  0.9302232589136954  0.5764152215010026                      0.0\n",
       "3     SNP_A-1861214    rs492861           G           A           G           A     0.9079122340425532     0.9092046023011505      1.015677563895  0.7207533541801461                      0.0\n",
       "4     SNP_A-1860431    rs496723           G           A           G           A     0.4557845744680851    0.45947973986993496  1.0149989992341526   0.999542988917936    2.052818933652854e-22\n",
       "5     SNP_A-1956539    rs510688           C           G           C           G      0.996343085106383     0.9979989994997499  1.8305805805805806  0.5771027456346043                      0.0\n",
       "6     SNP_A-4255384    rs518058           C           A           C           A    0.46708776595744683     0.4652326163081541   0.992572963723705  0.9997671733953388   2.2118958676356859e-13\n",
       "7     SNP_A-1963203    rs520764           G          -1           G           A                    1.0     0.9989994997498749                 0.0   0.573236918288303                      nan\n",
       "8     SNP_A-2178414    rs524965           C           G           C           G     0.3703457446808511    0.37018509254627313  0.9993112427149301  0.9873774881029727    2.55662516087757e-199\n",
       "9     SNP_A-2000312    rs532419           G           A           G           A    0.46043882978723405    0.46823411705852924  1.0318375553148063  0.9997229741263883    1.907260458390498e-18\n",
       "10    SNP_A-4285953    rs573378           C           T           C           T    0.24900265957446807    0.24687343671835918   0.988645992730704  0.9170662006167895       0.8094762728412335\n",
       "11    SNP_A-2062285    rs580771           G           A           G           A     0.7101063829787234     0.6935967983991996  0.9241213788886341   0.954970332189847                      0.0\n",
       "12    SNP_A-4284705    rs597479           C           T           C           T  0.0003324468085106383  0.0007503751875937969   2.258072590738423  0.5733056062430879       0.9897099256674965\n",
       "13    SNP_A-4260954    rs609031           G           A           G           A     0.7752659574468085     0.7661330665332666    0.94962805331083  0.8995402824397165                      0.0\n",
       "14    SNP_A-2073413    rs624969           G           A           G           A     0.9268617021276596     0.9189594797398699  0.8947960394636626  0.6982957566932599                      0.0\n",
       "15    SNP_A-4292038    rs666111           C          -1           C           G                    1.0     0.9994997498749375                 0.0  0.5728219156239178                      nan\n",
       "16    SNP_A-2253696    rs783883           G           A           G           A     0.7952127659574468     0.8126563281640821   1.117088113024724  0.8638481072473968                      0.0\n",
       "17    SNP_A-2219063    rs841392           G          -1           C           G                    1.0  0.0007503751875937969                 0.0  0.5730294430626224                      nan\n",
       "18    SNP_A-2144578    rs864560           G           A           G           A     0.1535904255319149    0.15607803901950976  1.0191918336553785  0.8125915710165198                      0.0\n",
       "19    SNP_A-2190868    rs948620           C           T           C           T    0.24335106382978725    0.23936968484242122  0.9784906819909327  0.9110956219430866       0.1257407749820825\n",
       "20    SNP_A-4264596    rs962923           T           G           T           G     0.7662898936170213     0.7638819409704852  0.9866916063090555  0.9050340372238652       0.7569943543112758\n",
       "21    SNP_A-1786988   rs1359759           G          -1           G           A                    1.0     0.9997498749374687                 0.0  0.5726143360630414                      nan\n",
       "22    SNP_A-2275684   rs1409806           G           A           G           A               0.578125     0.5940470235117559  1.0678423340160863  0.9961803640375901   1.9425547092644167e-70\n",
       "23    SNP_A-1857432   rs1543594           C          -1           C           T                    1.0     0.9989994997498749                 0.0   0.573236918288303                      nan\n",
       "24    SNP_A-2105952   rs1551623           T           G           T           G    0.48570478723404253     0.4912456228114057  1.0224230517883186  0.9999899814742175       0.3113572840262145\n",
       "25    SNP_A-2213809   rs1551624           G           A           G           A     0.5575132978723404      0.545272636318159  0.9517165192965331  0.9991710115097504     3.57792661570284e-39\n",
       "26    SNP_A-2118217   rs1695824           C           G           C           G    0.46675531914893614     0.4627313656828414  0.9839538005273574  0.9997357617006344  2.0239515589746665e-232\n",
       "27    SNP_A-4262266   rs1698597           G           A           G           A     0.5428856382978723     0.5327663831915957  0.9601061353691462  0.9996666807695281    1.915509712377738e-21\n",
       "28    SNP_A-2030917   rs1778121           C           G           C           G      0.840093085106383     0.8529264632316158   1.103867348284783   0.810716238260011   5.158118023773139e-292\n",
       "29    SNP_A-4253178   rs1835283           G           A           G           A     0.6615691489361702     0.6745872936468235  1.0604699129776476  0.9732289082335358                      0.0\n",
       "...             ...         ...         ...         ...         ...         ...                    ...                    ...                 ...                 ...                      ...\n",
       "2170  SNP_A-2300067  rs41517345           C          -1           C           T                    1.0     0.9979989994997499                 0.0  0.5740662952450257                      nan\n",
       "2171  SNP_A-1928576  rs41529146           C           T           C           T     0.6944813829787234     0.7106053026513257   1.080226796079228  0.9545414087070017       0.4377155488291514\n",
       "2172  SNP_A-4238377  rs41529846           C           T           C           T     0.6781914893617021     0.6835917958979489  1.0251662404092072  0.9670392672308434      0.14798543581212498\n",
       "2173  SNP_A-4280966  rs41531149           T           A           T           A    0.10871010638297872    0.11530765382691345  1.0685994022128673  0.7519530577485716                      0.0\n",
       "2174  SNP_A-2314782    rs140378           G           A           G           A     0.7928856382978723     0.7781390695347674  0.9161699744031538  0.8842983002519721                      0.0\n",
       "2175  SNP_A-4290489   rs5748616           G           A           G           A     0.3394281914893617     0.3129064532266133   0.886279645464895  0.9703401170148667                      0.0\n",
       "2176  SNP_A-2151351  rs41439550           C           T           C           T     0.6509308510638298     0.6558279139569785  1.0218587429032948  0.9794635080956823      0.13234237394837803\n",
       "2177  SNP_A-4284341   rs4010550           C           T           C           T     0.6894946808510638     0.6835917958979489  0.9729426248565908  0.9640279235175554    0.0056153861081748215\n",
       "2178  SNP_A-1854151   rs5747360           C           G           C           G     0.6891622340425532     0.6880940470235117  0.9950306205225392  0.9628896422777566  1.2680434459713623e-286\n",
       "2179  SNP_A-2276339   rs2379981           C           T           C           T     0.9820478723404256     0.9814907453726863  0.9693498508664385  0.6026503005703618      0.45104726180714083\n",
       "2180  SNP_A-4285196   rs4535153           G           A           G           A      0.854720744680851     0.8716858429214608   1.154688332829134  0.7877271184095397                      0.0\n",
       "2181  SNP_A-4229849  rs11089243           C           G           C           G     0.0581781914893617    0.05502751375687844  0.9426907660893897  0.6655664382423445    7.788837202109299e-39\n",
       "2182  SNP_A-2113726   rs5746647           C           G           C           G  0.0006648936170212766  0.0005002501250625312  0.7522522522522522  0.5733740121353723      0.15736831673292917\n",
       "2183  SNP_A-2214998   rs9605923           T           G           T           G    0.22340425531914893     0.2221110555277639  0.9925585668350941  0.8930107734743976      0.05452673337548858\n",
       "2184  SNP_A-2296864   rs5747999           C           G           C           G     0.5199468085106383     0.5500250125062531  1.1285596288930053  0.9996688083013009   2.890087559459401e-245\n",
       "2185  SNP_A-2145794  rs11089263           C           T           C           T    0.14760638297872342    0.13731865932966483  0.9192088580317053  0.7957168575997363       0.7170055228220931\n",
       "2186  SNP_A-4295548  rs11089264           C           T           C           T    0.08078457446808511    0.07578789394697348  0.9330760620792194  0.7002014465314452       0.7770893626472968\n",
       "2187  SNP_A-2256115   rs2154615           C           T           C           T      0.991688829787234     0.9947473736868434  1.5871685583385216  0.5836622508744744                      0.0\n",
       "2188  SNP_A-1813743   rs5993628           C           G           C           G     0.1612367021276596    0.14907453726863432    0.91135475190458  0.8129520473824803   3.0506929774379464e-99\n",
       "2189  SNP_A-1792553    rs361995           T           G           T           G     0.2237367021276596    0.23386693346673337  1.0590986994754534  0.8990387364709654       0.6871458183840945\n",
       "2190  SNP_A-2049082  rs12106650           T          -1           T           G                    1.0     0.9997498749374687                 0.0  0.5726143360630414                      nan\n",
       "2191  SNP_A-2287278   rs2845362           G           A           G           A     0.3796542553191489    0.38594297148574286   1.026975221232625  0.9906149141919861  1.6049536406440105e-178\n",
       "2192  SNP_A-1860833   rs2845377           G           A           G           A     0.3617021276595745    0.35542771385692845  0.9730877217010203  0.9837564537908076   2.277661307635628e-241\n",
       "2193  SNP_A-2027961   rs2215075           G           A           G           A    0.05053191489361702    0.05127563781890945  1.0155133417514257  0.6563325720104201                      0.0\n",
       "2194  SNP_A-1970657  rs17432784           G           A           G           A     0.0425531914893617    0.04452226113056528  1.0484293193717278  0.6443242463857755                      0.0\n",
       "2195  SNP_A-1936806  rs16984825           G           A           G           A     0.9996675531914894     0.9997498749374687  1.3292317924842034  0.5728903614495753                      0.0\n",
       "2196  SNP_A-2290590   rs8139954           C           A           C           A   0.041555851063829786    0.04302151075537769   1.036855201254574  0.6422823228744252                      0.0\n",
       "2197  SNP_A-4198761   rs5748617           G           A           G           A  0.0003324468085106383  0.0005002501250625312   1.505005005005005  0.5730980099333562                      0.0\n",
       "2198  SNP_A-2078648   rs9605146           C           T           C           T    0.10272606382978723    0.09954977488744372  0.9656616325062927  0.7356143142155831        0.600652738017438\n",
       "2199  SNP_A-1827561  rs16981694           C           T           C           T    0.11402925531914894    0.11480740370185093  1.0077091830556144  0.7555508549056137        0.057437616731301\n",
       "\n",
       "[2200 rows x 11 columns]"
      ]
     },
     "execution_count": 392,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1.,  1.,  1., ..., 22., 22., 22.])"
      ]
     },
     "execution_count": 389,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr = np.array([])\n",
    "for i in range(22):\n",
    "    i=i+1\n",
    "    for j in range(100):\n",
    "        arr=np.append(arr,i)\n",
    "arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "metadata": {},
   "outputs": [],
   "source": [
    "chro = pd.DataFrame(arr)\n",
    "a.insert(0,'chr',chro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-2272612  rs4247179\n",
      "1  SNP_A-2289056  rs4675917\n",
      "2  SNP_A-4199218  rs4675920\n",
      "3  SNP_A-2308630  rs6437265\n",
      "4  SNP_A-4292663  rs6437282\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name      rsid\n",
      "0  SNP_A-2039407  rs736480\n",
      "1  SNP_A-1972592  rs743660\n",
      "2  SNP_A-2275659  rs743661\n",
      "3  SNP_A-1974174  rs804968\n",
      "4  SNP_A-1974175  rs804974\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-1977622  rs1246399\n",
      "1  SNP_A-1883534  rs1688141\n",
      "2  SNP_A-1906396  rs1691263\n",
      "3  SNP_A-4198617  rs1908421\n",
      "4  SNP_A-1828242  rs2643003\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name      rsid\n",
      "0  SNP_A-1903590   rs26225\n",
      "1  SNP_A-2142466   rs26228\n",
      "2  SNP_A-2050841   rs29654\n",
      "3  SNP_A-4287143  rs156755\n",
      "4  SNP_A-1982036  rs156756\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name    rsid\n",
      "0  SNP_A-1930054  rs3077\n",
      "1  SNP_A-2172053  rs6531\n",
      "2  SNP_A-1894737  rs7194\n",
      "3  SNP_A-2125247  rs7565\n",
      "4  SNP_A-4195037  rs8233\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name      rsid\n",
      "0  SNP_A-2192884  rs378441\n",
      "1  SNP_A-2071014  rs382489\n",
      "2  SNP_A-4199336  rs411838\n",
      "3  SNP_A-2265133  rs413897\n",
      "4  SNP_A-2220591  rs427619\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-2067518   rs316472\n",
      "1  SNP_A-2295467  rs1464999\n",
      "2  SNP_A-4231325  rs1548128\n",
      "3  SNP_A-1852142  rs1907457\n",
      "4  SNP_A-4230826  rs2222675\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-2014238   rs2874557\n",
      "1  SNP_A-2062715   rs7863531\n",
      "2  SNP_A-1874310  rs35336397\n",
      "3  SNP_A-1859290  rs41318950\n",
      "4  SNP_A-1996140  rs41319051\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-1831478  rs17343324\n",
      "1  SNP_A-4245017  rs34243358\n",
      "2  SNP_A-2026252  rs41321150\n",
      "3  SNP_A-4207648  rs41321644\n",
      "4  SNP_A-2193246  rs41323348\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-4243770  rs2111401\n",
      "1  SNP_A-2011868  rs3867286\n",
      "2  SNP_A-1904484  rs3867288\n",
      "3  SNP_A-2042969  rs7396835\n",
      "4  SNP_A-2069009  rs7396851\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-4289334  rs11247027\n",
      "1  SNP_A-2300162  rs17879282\n",
      "2  SNP_A-1878197  rs41319045\n",
      "3  SNP_A-2037950  rs41319144\n",
      "4  SNP_A-2139134  rs41319846\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-1782699   rs4349050\n",
      "1  SNP_A-4303595   rs9581956\n",
      "2  SNP_A-2180321  rs41324245\n",
      "3  SNP_A-2130269  rs41326249\n",
      "4  SNP_A-2191692  rs41326444\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-1858314  rs41323047\n",
      "1  SNP_A-4233168  rs41324646\n",
      "2  SNP_A-2091422  rs41325346\n",
      "3  SNP_A-4196769  rs41325945\n",
      "4  SNP_A-4293737  rs41336446\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-2186013   rs6606854\n",
      "1  SNP_A-2291060   rs7172750\n",
      "2  SNP_A-2213005  rs12443222\n",
      "3  SNP_A-4216646  rs17561681\n",
      "4  SNP_A-4291787  rs41318651\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-2208239   rs235633\n",
      "1  SNP_A-4259333  rs7189888\n",
      "2  SNP_A-2097341  rs7190132\n",
      "3  SNP_A-2298053  rs7191613\n",
      "4  SNP_A-1796741  rs7193756\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-1822315   rs296142\n",
      "1  SNP_A-2193198   rs365424\n",
      "2  SNP_A-4214460   rs434005\n",
      "3  SNP_A-4203177   rs439075\n",
      "4  SNP_A-2098700  rs1106175\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-4285539  rs41320644\n",
      "1  SNP_A-4285674  rs41325953\n",
      "2  SNP_A-2049475  rs41326951\n",
      "3  SNP_A-2203015  rs41329244\n",
      "4  SNP_A-4299856  rs41330145\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-2287868    rs607149\n",
      "1  SNP_A-2124700    rs613238\n",
      "2  SNP_A-4245521    rs618835\n",
      "3  SNP_A-4207784    rs629122\n",
      "4  SNP_A-1961755  rs11555837\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name        rsid\n",
      "0  SNP_A-2135418  rs41324045\n",
      "1  SNP_A-1936599  rs41325547\n",
      "2  SNP_A-2311143  rs41340548\n",
      "3  SNP_A-4228972  rs41346144\n",
      "4  SNP_A-1793035  rs41352444\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-2020296  rs2838224\n",
      "1  SNP_A-4270683  rs2838227\n",
      "2  SNP_A-1970556  rs2838243\n",
      "3  SNP_A-4263122  rs2838246\n",
      "4  SNP_A-4270685  rs8127663\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n",
      "        snp_name       rsid\n",
      "0  SNP_A-4257236   rs134871\n",
      "1  SNP_A-4272387  rs2223273\n",
      "2  SNP_A-4282160  rs5751251\n",
      "3  SNP_A-2290765  rs5758659\n",
      "4  SNP_A-1910486  rs5758660\n",
      "        snp_name      rsid\n",
      "0  SNP_A-4221087  rs307378\n",
      "1  SNP_A-2203152  rs474731\n",
      "2  SNP_A-4198280  rs485777\n",
      "3  SNP_A-1861214  rs492861\n",
      "4  SNP_A-1860431  rs496723\n"
     ]
    }
   ],
   "source": [
    "snpsid = pd.DataFrame(columns = [\"snp_name\",\"rsid\"])\n",
    "for i in range(22):\n",
    "    i = i+1\n",
    "    \n",
    "    if i<10:\n",
    "        stro = '0'+str(i)\n",
    "    else:\n",
    "        stro = str(i)\n",
    "    snp = pd.read_csv('BWSI_100set/58C/snps/snps_'+stro,header=None, sep='\\t')\n",
    "    \n",
    "    snp = snp.drop([0,1],axis=1)\n",
    "    snp = snp.drop([2],axis=1)\n",
    "    \n",
    "    snp = snp.rename(columns={3:'snp_name',4:'rsid'} )\n",
    "    print(snp.head())\n",
    "    snpsid = snpsid.append(snp)\n",
    "    print(snpsid.head())\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(22):\n",
    "    i=i+1\n",
    "    if i<10:\n",
    "        stro = '0' + str(i)\n",
    "    else:\n",
    "        stro = str(i)\n",
    "    snp = pd.read_csv('BWSI_100set/58C/snps/snps_'+stro,header=None, sep='\\t')\n",
    "    for j in range(100):\n",
    "        snpid.iloc[j,1] = snp.iloc[j,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "bigdf = bigdf.rename(columns={0:'Con Minor Allele',1:'Con Major Allele',2:'Dis Minor Allele',3:'Dis Major Allele',4:'Con Minor Allele Freq',5:'Dis Minor Allele Freq',6:'Odds Ratio',7:'P value',8:'HWE'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Odds Ratio:\n",
      "0.9963249409766973\n",
      "\n",
      "Type 2 Diabetes:\n",
      "0.38244122061030517\n",
      "0.8141174907213498\n",
      "0.8154681593364395\n",
      "0.6175587793896948\n",
      "\n",
      "Control:\n",
      "0.38331117021276595\n",
      "0.616688829787234\n",
      "\n",
      "Chi-Squared:\n",
      "0.9906428272745443\n",
      "\n",
      "HWE:\n",
      "0.9149630590292988\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1153.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1855.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1529.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2469.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.383311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.382441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.996325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.990643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.914963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0\n",
       "0  1153.000000\n",
       "1  1855.000000\n",
       "2  1529.000000\n",
       "3  2469.000000\n",
       "4     0.383311\n",
       "5     0.382441\n",
       "6     0.996325\n",
       "7     0.990643\n",
       "8     0.914963"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Odds Ratio:\")\n",
    "print(str(computeOR(getAlleleCounts(c1_1),getAlleleCounts(t2d_1))))\n",
    "\n",
    "print(\"\\nType 2 Diabetes:\")\n",
    "print(computeAlleleFrequency(getAlleleCounts(t2d_1)[0], getAlleleCounts(t2d_1)[1]))\n",
    "\n",
    "t1_minor = computeAlleleFrequency(getAlleleCounts(t2d_1)[0], getAlleleCounts(t2d_1)[1])\n",
    "t1_major = 1 - t1_minor\n",
    "control_minor = computeAlleleFrequency(getAlleleCounts(c1_1)[0], getAlleleCounts(c1_1)[1])\n",
    "control_major = 1-control_minor\n",
    "\n",
    "print(getPValue([t1_minor, t1_major]))\n",
    "print(getPValue([control_minor, control_major]))\n",
    "\n",
    "print(1-computeAlleleFrequency(getAlleleCounts(t2d_1)[0], getAlleleCounts(t2d_1)[1]))\n",
    "print(\"\\nControl:\")\n",
    "print(computeAlleleFrequency(getAlleleCounts(c1_1)[0], getAlleleCounts(c1_1)[1]))\n",
    "\n",
    "print(1-computeAlleleFrequency(getAlleleCounts(c1_1)[0], getAlleleCounts(c1_1)[1]))\n",
    "\n",
    "# print(control)\n",
    "\n",
    "print(\"\\nChi-Squared:\")\n",
    "print(getPValue([t1_minor, t1_major,control_minor, control_major]))\n",
    "\n",
    "print(\"\\nHWE:\")\n",
    "print(HWEChiSq(control_minor, [c1_1.to_string().count('C C'), c1_1.to_string().count('C T'), c1_1.to_string().count('T T')]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp = np.array([])\n",
    "for i in range(22):\n",
    "    i = i+1\n",
    "    print(\"Chromosome \"+ str(i)+\"...\")\n",
    "    if i<10:\n",
    "        stro = '0'+str(i)\n",
    "    else:\n",
    "        stro=str(i)\n",
    "    dat = pd.read_csv('BWSI_100set/58C/snps/snps_'+stro,header=None, sep='\\t')\n",
    "    dat = dat.drop([0,1],axis=1)\n",
    "    snpid = dat.loc[:,3]\n",
    "    for j in snpid:\n",
    "        snp=np.append(snp,j)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SNP_A-2203152', 'SNP_A-4198280'], dtype='<U32')"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22, 41400)"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SNP_A-4221087\n",
       "1    SNP_A-2203152\n",
       "2    SNP_A-4198280\n",
       "3    SNP_A-1861214\n",
       "4    SNP_A-1860431\n",
       "Name: 3, dtype: object"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat = pd.read_csv('BWSI_100set/58C/snps/snps_01',header=None, sep='\\t')\n",
    "dat = dat.drop([0,1],axis=1)\n",
    "dat.head()\n",
    "snpid = dat.loc[:,3]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hardy Weinberg Equilibrium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function takes as input the minor allele frequency (p) and the population genotype counts\n",
    "## returns the p-value from the chisq test\n",
    "def HWEChiSq (p, populationGenotypeCounts):\n",
    "    \"\"\"\n",
    "        |  CC | CT  | TT |\n",
    "    BP  | 270 | 957 | 771|\n",
    "    compute the HWE\n",
    "      -- p = frequency of the C allele\n",
    "      -- q = 1-p = frequency of the T allele\n",
    "      -- if the population genotype is in Hardy Weinberg Equilibrium, we would expect the genotype frequencies to be\n",
    "          CC = p^2*N \n",
    "          CT = 2pq*N \n",
    "          TT = q^2*N \n",
    "      -- to compute the deviation from HWE\n",
    "         (observed - expected)^2 / expected\n",
    "      -- do a chi.squared test to check if the deviation is significant\n",
    "    \"\"\"\n",
    "    total = sum(populationGenotypeCounts)\n",
    "        \n",
    "    q = 1-p\n",
    "    \n",
    "    observed = populationGenotypeCounts[0]\n",
    "    expected = p**2*total\n",
    "    \n",
    "    CC = (observed-expected)**2/expected\n",
    "    \n",
    "    \n",
    "    observed = populationGenotypeCounts[1]\n",
    "    expected = 2*p*q*total\n",
    "    \n",
    "    CT = (observed-expected)**2/expected\n",
    "    \n",
    "    observed = populationGenotypeCounts[2]\n",
    "    expected = q**2*total\n",
    "    \n",
    "    TT = (observed-expected)**2/expected\n",
    "    \n",
    "    from scipy.stats import chisquare\n",
    "    from scipy.stats.distributions import chi2\n",
    "    return chi2.sf(CC+CT+TT, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {},
   "outputs": [],
   "source": [
    "## plot the data\n",
    "## manhattan plot\n",
    "## the X axis represents the chromosomes \n",
    "## the y-axis is the -log(p.value)\n",
    "def manhattanPlot ():\n",
    "    \"\"\"\n",
    "    group data by chromosomes, SNP before plotting\n",
    "    \"\"\"\n",
    "    pd.set_option('expand_frame_repr', False)\n",
    "    df = pd.read_csv('gene_dat.csv')\n",
    "    \n",
    "\n",
    "    df['p_adj'] = -np.log10(df.p)\n",
    "    df.chr = df.chr.astype('category')\n",
    "\n",
    "    # print(df.head(10))\n",
    "\n",
    "    df['ind'] = range(len(df))\n",
    "    df_grouped = df.groupby(('chr'))\n",
    "\n",
    "    # print(df_grouped.head(10))\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111)\n",
    "    colors = ['#E24E42','#008F95']\n",
    "    x_labels = []\n",
    "    x_labels_pos = []\n",
    "\n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        group.plot(kind='scatter', x='ind', y='p_adj',color=colors[num % len(colors)], ax=ax, s=5)\n",
    "        x_labels.append(name)\n",
    "        x_labels_pos.append((group['ind'].iloc[-1] - (group['ind'].iloc[-1] - group['ind'].iloc[0])/2))\n",
    "\n",
    "    ax.set_xticks(x_labels_pos)\n",
    "    ax.set_xticklabels(x_labels)\n",
    "    ax.set_xlim([0, len(df)])\n",
    "    ax.set_ylim([0, 6])\n",
    "    ax.set_xlabel('Chromosome')\n",
    "    plt.xticks(fontsize = 8,rotation=60)\n",
    "    plt.yticks(fontsize = 8)\n",
    "\n",
    "    # xticks = ax.xaxis.get_major_ticks()\n",
    "    # xticks[0].set_visible(False)\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>snp_name</th>\n",
       "      <th>rs</th>\n",
       "      <th>con min all</th>\n",
       "      <th>dis min all</th>\n",
       "      <th>con maj all</th>\n",
       "      <th>dis maj all</th>\n",
       "      <th>con maj all freq</th>\n",
       "      <th>dis maj all freq</th>\n",
       "      <th>odds rat</th>\n",
       "      <th>p</th>\n",
       "      <th>HWE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>SNP_A-4221087</td>\n",
       "      <td>rs307378</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.530918</td>\n",
       "      <td>0.558279</td>\n",
       "      <td>1.116672</td>\n",
       "      <td>0.999392</td>\n",
       "      <td>9.226596e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SNP_A-2203152</td>\n",
       "      <td>rs474731</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.349069</td>\n",
       "      <td>0.336168</td>\n",
       "      <td>0.944326</td>\n",
       "      <td>0.977834</td>\n",
       "      <td>1.482832e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>SNP_A-4198280</td>\n",
       "      <td>rs485777</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.997673</td>\n",
       "      <td>0.997499</td>\n",
       "      <td>0.930223</td>\n",
       "      <td>0.576415</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>SNP_A-1861214</td>\n",
       "      <td>rs492861</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.907912</td>\n",
       "      <td>0.909205</td>\n",
       "      <td>1.015678</td>\n",
       "      <td>0.720753</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>SNP_A-1860431</td>\n",
       "      <td>rs496723</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.455785</td>\n",
       "      <td>0.459480</td>\n",
       "      <td>1.014999</td>\n",
       "      <td>0.999543</td>\n",
       "      <td>2.052819e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>SNP_A-1956539</td>\n",
       "      <td>rs510688</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.996343</td>\n",
       "      <td>0.997999</td>\n",
       "      <td>1.830581</td>\n",
       "      <td>0.577103</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>SNP_A-4255384</td>\n",
       "      <td>rs518058</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.467088</td>\n",
       "      <td>0.465233</td>\n",
       "      <td>0.992573</td>\n",
       "      <td>0.999767</td>\n",
       "      <td>2.211896e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>SNP_A-1963203</td>\n",
       "      <td>rs520764</td>\n",
       "      <td>G</td>\n",
       "      <td>-1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.998999</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.573237</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>SNP_A-2178414</td>\n",
       "      <td>rs524965</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.370346</td>\n",
       "      <td>0.370185</td>\n",
       "      <td>0.999311</td>\n",
       "      <td>0.987377</td>\n",
       "      <td>2.556625e-199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>SNP_A-2000312</td>\n",
       "      <td>rs532419</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.460439</td>\n",
       "      <td>0.468234</td>\n",
       "      <td>1.031838</td>\n",
       "      <td>0.999723</td>\n",
       "      <td>1.907260e-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0       snp_name        rs con min all dis min all con maj all dis maj all  con maj all freq  dis maj all freq  odds rat         p            HWE\n",
       "0           0  SNP_A-4221087  rs307378           C           T           C           T          0.530918          0.558279  1.116672  0.999392   9.226596e-01\n",
       "1           1  SNP_A-2203152  rs474731           C           T           C           T          0.349069          0.336168  0.944326  0.977834   1.482832e-01\n",
       "2           2  SNP_A-4198280  rs485777           T           A           T           A          0.997673          0.997499  0.930223  0.576415   0.000000e+00\n",
       "3           3  SNP_A-1861214  rs492861           G           A           G           A          0.907912          0.909205  1.015678  0.720753   0.000000e+00\n",
       "4           4  SNP_A-1860431  rs496723           G           A           G           A          0.455785          0.459480  1.014999  0.999543   2.052819e-22\n",
       "5           5  SNP_A-1956539  rs510688           C           G           C           G          0.996343          0.997999  1.830581  0.577103   0.000000e+00\n",
       "6           6  SNP_A-4255384  rs518058           C           A           C           A          0.467088          0.465233  0.992573  0.999767   2.211896e-13\n",
       "7           7  SNP_A-1963203  rs520764           G          -1           G           A          1.000000          0.998999  0.000000  0.573237            NaN\n",
       "8           8  SNP_A-2178414  rs524965           C           G           C           G          0.370346          0.370185  0.999311  0.987377  2.556625e-199\n",
       "9           9  SNP_A-2000312  rs532419           G           A           G           A          0.460439          0.468234  1.031838  0.999723   1.907260e-18"
      ]
     },
     "execution_count": 385,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Write code to run the GWAS\n",
    "\n",
    "## Write a function that will take the following inputs:\n",
    "## (a) control files - 58C and NBS\n",
    "## (b) the disease - T2D, T1D, HT, BD, CD, RA\n",
    "## (c) the chromosome number\n",
    "\n",
    "## Your function should output a csv file with the following information\n",
    "## SNP Id, RSID, CHROMOSOME NUMBER, MINOR ALLELE, MAJOR ALLELE, MINOR ALLELE FREQUENCY, MAJOR ALLELE FREQUENCY, \n",
    "## ODDS RATIO (MINOR vs MAJOR ALLELE), P-Value for ODDS RATIO, Chi square for the ODDS RATIO, HWE deviation P-VALUE\n",
    "## \n",
    "\n",
    "def gwas (controlFile, disease, chrom_num):\n",
    "    \"\"\"\n",
    "    1. Load the control file for the specified chromosome number. Since there are two control populations, \n",
    "       it would be best to combine the two controls together\n",
    "    2. Load the disease file for the given chromosome number\n",
    "    3. Split the header from the SNP data for the controls and the disease\n",
    "    3. Load the SNP file -- this gives you the mapping between the WTCCC SNP ids and the RSID\n",
    "    4. For each SNP in the disease file and its matching SNP in the control file:\n",
    "       -- to make it easier, before you begin do a check for zygosity. If either the controls or the disease \n",
    "       are homozygous, skip the SNP\n",
    "       genotype CC, TT : skip\n",
    "       genotype CC CT TT : run the following analysis\n",
    "       a. get the minor and major allele counts and compute the minor allele frequency for the disease population\n",
    "       b. get the minor and major allele counts and compute the minor allele frequency for the control population\n",
    "       c. conduct the allelic test -- compute the odds ratios to test the frequency of the minor allele \n",
    "          in the disease population compared to the control population\n",
    "          -- you can build a confusion matrix based on the counts that you have calculated in a and b\n",
    "          -- compute your odds ratios based on the minor allele of the control population\n",
    "          \n",
    "          |                  | C    | T    | \n",
    "          |------------------+ ---  + ---  +\n",
    "          | Bipolar Disorder | 1529 | 2469 |\n",
    "          |------------------+----  + ---  +\n",
    "          | Healthy Controls | 2270 | 3738 |\n",
    "          |------------------+----- +----- +\n",
    "          \n",
    "          (Test : your OR for above should be :1.019)\n",
    "       d. Do a chi square test to test the significant and record the p-value\n",
    "       d. for each SNP calculate the deviation from Hardy-Weinberg equilibrium\n",
    "       e. create an output row containing:\n",
    "       SNP Id, RSID, CHROMOSOME NUMBER, MINOR ALLELE, MAJOR ALLELE, MINOR ALLELE FREQUENCY, \n",
    "       MAJOR ALLELE FREQUENCY, ODDS RATIO (MINOR vs MAJOR ALLELE), \n",
    "       P-Value for ODDS RATIO, CHI_SQ for the ODDS RATIO, HWE deviation P-VALUE  \n",
    "    5. Save the result file for the chromosome\n",
    "    \"\"\"\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Write a main function that will invoke gwas for all chromosomes (1 through 22, don't include the X chromosome)\n",
    "\n",
    "def runGWAS (disease):\n",
    "    \"\"\"\n",
    "    1. For each chromosome 1:22 call the gwas function\n",
    "    2. Merge all the result files into one \n",
    "    3. Filter your data\n",
    "       a. remove all SNPs that deviate signficantly from HWE equilibrium -- these may indicate an error or \n",
    "          population specific deviation. For this remove all SNPs for which HWE deviation P-Value < 0.05\n",
    "       b. remove all SNPs that have minor allele frequency less than 1% -- not enough data\n",
    "    4. Draw a manhattan plot\n",
    "       c. draw a line at your Bonferroni threshold (Bonferroni correction): \n",
    "       if your significance is set at p.value < 0.05, then Bonferroni correction = 0.05/number of tests that you ran\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"manhattan.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## How many SNPs did you identify that exceeded the Bonferroni-level of significance?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How much can you trust your results\n",
    "\n",
    "Typically, you want to examine the genome-wide distribution of your chi_square distribution against an expected null distribution. Visually you can do this using a quantile-quantile plot. Significant deviation from the null distribution may point to population stratification, familial relationships, technical bias, poor sample collection or sometimes even undetected sample duplications.\n",
    "(Slide courtesy - Dr. Chirag Patel)\n",
    "\n",
    "<img src=\"qqplot_example.png\">\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Based on your results do a quantile-quantile plot for your data\n",
    "\n",
    "##load your results \n",
    "##sort/order the -log10 of your p values in increasing order\n",
    "##generate an expected set of -log10(p_values)\n",
    "##plot the expected values on the x-axis and the observed -log10 (pvalues) on the y axis\n",
    "##Alternatively use: p = observed p values\n",
    "##stats.probplot(p, dist=\"norm\", plot=pylab)\n",
    "##pylab.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genomic Inflation Factor\n",
    "\n",
    "Mathematically, you can correct for some of the deviation by computing an inflation factor and then correcting for this inflation factor. This quantity is termed as genomic inflation factor ($\\lambda$)\n",
    "\n",
    "$\\lambda$ = observed median of test statistic distribution / expected median of the test statistic \n",
    "distribution.  For 1 degree of freedom, the X2 distribution has an expected median of 0.455\n",
    "\n",
    "$\\lambda$ <= 1.05 is considered acceptable. >1.1 is troubling, and indicates there is some inflation of the p values.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compute the genomic inflation factor\n",
    "\n",
    "##labmda =  median(CHI2)/qchisq(0.5, 1)\n",
    "##if (gif > 1.05), correct your computed Chi.square values and recompute the p_values\n",
    "##chi2_corrected = CHI2/lambda, \n",
    "##pval_corrected = pchisq(chi2_corrected, 1, lower.tail = F)\n",
    "##re-draw the qqplot using the corrected pvalues (pval_corrected)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What does this mean?\n",
    "\n",
    "Identify a SNP that is significant (above the Bonferroni threshold). \n",
    "What is the p-value of association and odds ratio for the SNP?\n",
    "What is the interpretation of the odds ratio? \n",
    "What gene is this SNP associated with and its putative function in your disease pathogenesis? \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Polygenic Risk Scores (PRS)\n",
    "\n",
    "A PRS is the cumulative risk of all SNP for a single patient. For example, if a patient has a 2 snps with OR=2 each, then their PRS is 4. The underlying assumption here is that each allele has an equal contribution and there is no interation between SNPS, that is the SNPS are independent. \n",
    "\n",
    "$prs =\\sum_{i=1}^{m}-log_{10}(x_i) \\cdot n_i$"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
